docs/design/proposals/codegen-refactor/model_cmd_2.md

```diff
@@ -0,0 +1,143 @@
+
+Current Pipeline
+---
+
+![current-pipeline](./images/current_pipeline.png)
+* Displayed are the important and mostly duplicated calls between the `apis` and `controller` commands
+* The cache is only used to store the `aws-sdk` repo
+* The red line denotes a dependency on $SERVICE-controller (output directory) repo that should be removed
+* Only `GetCRDs` is duplicated between the commands when creating the model
+
+
+Current Dev Loop
+---
+
+![current-dev-loop](./images/current_dev_loop.png)
+* Validating the result of a change in *code-generator* and/or *generator.yaml* can be a lengthy process
+* In the current feedback loop, users and contributors need to manually verify files or run lengthy e2e tests even for minor changes to `ackgenconfig`
+
+
+## Solution
+Add a new command, `model`, which takes in `aws-sdk` and *generator.yaml* as inputs to create and cache `ackmodel`. Existing commands, `api` and `controller`, will remove parameters for `aws-sdk` and *generator.yaml* and replace them with `ackmodel` input and/or let it default to `--cache-dir`.
+
+Updated Pipeline
+---
+![updated-pipeline](./images/proposed_pipeline.png)
+* This is not an exhaustive diagram of the calls, but shows the clear responsibility of `generateModel` and how downstream commands like `api` and `controller` become significantly lighter and easier to follow.
+* `ackmodel` will be cached in the same folder as `aws-sdk` and downstream commands will check for `ackmodel` instead of `aws-sdk`
+* users can override with a path to their own `ackmodel` (for even quicker testing!), but by default will check `--cach-dir`
+* No more dependency on $SERVICE-controller repo. The `ackmodel` will be generated with a specific version which can be extracted after commands `GetACKModel()`
+
+Updated Dev Loop:
+---
+![updated-dev-loop](./images/proposed_dev_loop.png)
+* Diffing the `ackmodel` between *code-generator* runs will be quicker than manually checking files or running e2e tests
+* Users can directly edit `ackmodel` or underlying `aws-sdk` / `ackgenconfig` then generate $SERVICE-controller for even quicker iteration loops
+
+
+Updated `ackmodel` creation:
+---
+To address the inconsistent `ackmodel` creation, I propose we edit `aws-sdk` directly as much as we can, then introduce more fields to represent other relations that can't be applied by editing `aws-sdk`. Therefore, if we want to rename a field/resource, it would be instead completely removed from model. Conceptually:
+
+```
+func (m *Model) RenameShapes() error {
+  for shapeName, shape := range m.SDKAPI.API.Shapes {
+    if m.cfg.shapeRenamed(shapeName) {
+      updateName := m.cfg.GetShapeName(shapeName)
+      m.SDKAPI.Shapes[updatedName] = shape
+      delete(shapeName, shape)
```

Comment:
I like this concept! Keep a modified version of the AWS SDK as our new base.

Unfortunately it's not quite as simple as this in reality. `ShapeName` is referenced both in the key of the `Shapes` map, but also in the `ShapeName` attribute of a `ShapeRef`. This means that updating the shape name would require finding all `ShapeRef` objects which match the shape and updating their attributes (this is non-trivial and the time complexity is recursive on number of attributes per shape).

Another concern I have is that we are now strongly coupled to the AWS SDK. We don't have plans to update to v2, but keeping some level of abstraction above the AWS SDK protects us from the subtle changes they could make to their internals.

Reply:
Agreed this would be a very complex endeavor, but you're saying it's possible?? :)  Is your concern feasibility or the performance hit and having to maintain the more complex code? This could also be an ongoing effort where configs slowly start to edit aws sdk over time.

> Another concern I have is that we are now strongly coupled to the AWS SDK. We don't have plans to update to v2, but keeping some level of abstraction above the AWS SDK protects us from the subtle changes they could make to their internals.

I see your point. I think moving methods that directly edit aws-sdk from `Model` to [SDK Helper](https://github.com/aws-controllers-k8s/code-generator/blob/main/pkg/sdk/helper.go#L50) and having an SDK Helper implemented for each supported api version could be a better abstraction.  This way we can support `v1.RenameShapes()` and `v2.RenameShapes()` if the sdks differed significantly. What do you think?

Reply:
> Agreed this would be a very complex endeavor, but you're saying it's possible

Tentatively yes. Doesn't mean we should be, though.